packages/tpetra/core/src/Tpetra_replaceDiagonalCrsMatrix_def.hpp

```diff
@@ -0,0 +1,150 @@
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Michael A. Heroux (maherou@sandia.gov)
+//
+// ************************************************************************
+// @HEADER
+
+#ifndef TPETRA_REPLACEDIAGONALCRSMATRIX_DEF_HPP
+#define TPETRA_REPLACEDIAGONALCRSMATRIX_DEF_HPP
+
+/// \file Tpetra_replaceDiagonalCrsMatrix_def.hpp
+/// \brief Definition of Tpetra::repalceDiagonalCrsMatrix
+
+#include "Tpetra_Details_Behavior.hpp"
+#include "Tpetra_CrsMatrix.hpp"
+#include "Tpetra_Vector.hpp"
+
+namespace Tpetra {
+
+template<class SC, class LO, class GO, class NT>
+LO
+replaceDiagonalCrsMatrix (::Tpetra::CrsMatrix<SC, LO, GO, NT>& matrix,
+    const ::Tpetra::Vector<SC, LO, GO, NT>& newDiag) {
+
+  using map_type = ::Tpetra::Map<LO, GO, NT>;
+  using crs_matrix_type = ::Tpetra::CrsMatrix<SC, LO, GO, NT>;
+  // using vec_type = ::Tpetra::Vector<SC, LO, GO, NT>;
+
+  const LO zeroLO = Teuchos::OrdinalTraits<LO>::zero();
+  const LO oneLO = Teuchos::OrdinalTraits<LO>::one();
+
+  // Count number of successfully replaced diagonal entries
+  LO numReplacedDiagEntries = 0;
+
+  // Extract some useful data
+  auto rowMapPtr = matrix.getRowMap();
+  if (rowMapPtr.is_null() || rowMapPtr->getComm().is_null()) {
+    // Processes on which the row Map or its communicator is null
+    // don't participate.  Users shouldn't even call this method on
+    // those processes.
+    return numReplacedDiagEntries;
+  }
+  auto colMapPtr = matrix.getColMap();
+
+  TEUCHOS_TEST_FOR_EXCEPTION
+    (colMapPtr.get () == nullptr, std::invalid_argument,
```

Comment:
It's perfectly fine to throw here.  The matrix needs a column Map in order to know what entries are diagonal entries.